pkg/controller/machine.go

```diff
@@ -394,48 +394,51 @@ func (c *Controller) deleteMachineAndProviderInstance(prov cloud.Provider, provi
 	}
 	glog.V(4).Infof("Deleted machine %s at cloud provider", machine.Name)
 
-	// step 3: remove finalizers this essentially will remove the machine object from the system
+	// step 3: remove error message in case it was set
```

Comment:
I moved `step 3` before `step 4` because it caused errors like this one:

` failed to update machine after removing the delete error: Operation cannot be fulfilled on machines.machine.k8s.io "machine-kubermatic-z5psw2ws48-m9hh5": StorageError: invalid object, Code: 4, Key: /registry/machine.k8s.io/machines/machine-kubermatic-z5psw2ws48-m9hh5, ResourceVersion: 0, AdditionalErrorMsg: Precondition failed: UID in precondition: 389cc5b7-3732-11e8-bdd4-0a580a2c021a, UID in object meta: `

removing finalizers can cause the object to be garbage collected.

Reply:
I noticed that error as well while implementing vsphere, the following patch fixed it:

```
commit 7f95e3bb920db0aea2c4ae5b7c5316435c0b121d
Author: Alvaro Aleman <mailaddr>
Date:   Sat Mar 31 17:18:55 2018 +0200

    Only remove machine error if set

diff --git a/pkg/controller/machine.go b/pkg/controller/machine.go
index d9e039e..8348079 100644
--- a/pkg/controller/machine.go
+++ b/pkg/controller/machine.go
@@ -227,9 +227,12 @@ func (c *Controller) updateMachine(machine *machinev1alpha1.Machine) (*machinev1
 }
 
 func (c *Controller) clearMachineErrorIfSet(machine *machinev1alpha1.Machine) (*machinev1alpha1.Machine, error) {
-       machine.Status.ErrorMessage = nil
-       machine.Status.ErrorReason = nil
-       return c.updateMachine(machine)
+       if machine.Status.ErrorMessage != nil || machine.Status.ErrorReason != nil {
+               machine.Status.ErrorMessage = nil
+               machine.Status.ErrorReason = nil
+               return c.updateMachine(machine)
+       }
+       return machine, nil
```

Reply:
in this case it might be a little bit different, since removing finalizers from the object can actually remove the object from the system.